pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+    else:
+        decoded = line
+
+    stripped = decoded.rstrip("\n")
+
+    if stripped != decoded:
+        stripped = stripped.rstrip("\r")
+
+    return len(stripped)
+
+
+def _map_positions_to_result(
+    line: _StrLike,
+    search_dict: Dict[_StrLike, _BadChar],
+    new_line: _StrLike,
+    byte_str_length: int = 1,
+) -> Dict[int, _BadChar]:
+    """Get all occurrences of search dict keys within line
+
+    Ignores Windows end of line and can handle bytes as well as string.
+    Also takes care of encodings for which the length of an encoded code point does not
+    default to 8 Bit.
+    """
+
+    result: Dict[int, _BadChar] = {}
+    ignore_pos: Optional[int]
+    for search_for, char in search_dict.items():
+        if search_for not in line:
+            continue
+
+        # Special Handling for Windows '\r\n'
+        if char.unescaped == "\r" and line.endswith(new_line):
+            ignore_pos = len(line) - 2 * byte_str_length
+        else:
+            ignore_pos = None
+
+        start: int = 0
+        pos: int = line.find(search_for, start)
+        while pos > 0:
+            if pos != ignore_pos:
+                # Calculate the
+                col = int(pos / byte_str_length)
+                result[col] = char
+            start = pos + 1
+            pos = line.find(search_for, start)
+
+    return result
+
+
+UNICODE_BOMS: Dict[str, bytes] = {
+    "utf-8": codecs.BOM_UTF8,
+    "utf-16": codecs.BOM_UTF16,
+    "utf-32": codecs.BOM_UTF32,
+    "utf-16le": codecs.BOM_UTF16_LE,
+    "utf-16be": codecs.BOM_UTF16_BE,
+    "utf-32le": codecs.BOM_UTF32_LE,
+    "utf-32be": codecs.BOM_UTF32_BE,
+}
+
+
+@lru_cache(maxsize=1)
+def get_bom_sorted() -> Dict[bytes, str]:
+    """Get all BOM sorted by size, biggest first"""
+    result = OrderedDict()
+    for codec, bom in sorted(
+        UNICODE_BOMS.items(), key=lambda x: len(x[1]), reverse=True
+    ):
+        if codec not in ("utf-16", "utf-32"):
+            # only use codecs with defined endian
+            result[bom] = codec
+    return result
+
+
+def _normalize_codec_name(codec: str) -> str:
+    """Make sure the codec name is always gives as defined in the BOM dict"""
+    codec = codec.lower()
+    return re.sub("utf[ -]?(8|16|32)[ -]?(le|be|)", r"utf-\1\2", codec)
+
+
+def _remove_bom(encoded: bytes, encoding: str) -> bytes:
+    """remove the bom if given from a line"""
+    encoding = _normalize_codec_name(encoding)
+    if not encoding.startswith("utf"):
+        return encoded
+    bom = UNICODE_BOMS[encoding]
+    if encoded.startswith(bom):
+        return encoded[len(bom) :]
+    return encoded
+
+
+def _encode_without_bom(string: str, encoding: str) -> bytes:
+    """Encode a string but remove the BOM"""
+    return _remove_bom(string.encode(encoding), encoding)
+
+
+def _byte_to_str_length(codec: str) -> int:
+    """return how many byte are usually(!) a character point"""
+    codec = codec.replace("-", "").replace(" ", "").lower()
```

Comment:
Is this necessary? I think only the call to `.lower()` is necessary, all `codecs` passed to `_byte_to_str_length` seem to have `-` and the space doesn't matter for `startswith`.

We could even optimise and do `"32" in codec` but I can imagine that might become problematic when we encounter non-utf codecs?

Reply:
I guess "32" in codec is not a good idea indeed.
I just use `_normalize_codec_name` now and added a simple unit test.

Reply:
codec is normalized before now.